### ISSUE TITLE
Handle binders and kernels dying

### ISSUE THREAD

**Issue by akhmerov** [executablebooks/thebe#255]
As far as I understand, thebe right now doesn't detect when kernels or binders stop being available because of culling for instance. Catching the proper errors and requesting kernels or binders again would be a way to improve on that.

**Comment by akhmerov** [comment]
Indeed, since binder kills kernels really fast, the users are likely to face thebe UI breaking with this in the console:

![image](https://user-images.githubusercontent.com/2069677/96335240-9c0a3780-1077-11eb-91c4-997416615bc5.png)

**Comment by Miniland1333** [comment]
When the kernel or binder dies, would it be acceptable to force the user to refresh their page through a notification before requesting a new one? An issue if it is restarted transparently/automatically is that state is lost and this ambiguity may be confusing to the users. Alternatively, a reconnect can flush all Thebe output but this would have to be done in a way that doesn't startle the user.

Hypothetical example:
User has 10 code blocks that need to be ran sequentially. While after running block 5, they wait long enough to timeout. If the kernel or binder is refreshed automatically, running block 6 would likely fail as none of the prerequisite code in blocks 1-5 had been ran in the new kernel. However, the Thebe output currently does not change so unless the page is refreshed or we code something in there would be a disconnect between the visual and what is happening in the kernel.

**Comment by akhmerov** [comment]
I think it's reasonable if thebe shows an error message when it detects that the kernel died, and then restarts automatically. This is also similar to the jupyter behavior.

**Comment by Miniland1333** [comment]
When this happens, does Jupyter automatically clear the output or in the scenario I outlined above would it throw an error?

Since at LibreTexts a lot of our users are using Thebe as their first exposure to Jupyter (and may not even know what Jupyter is under the hood), I want to ensure this is as clear as possible for them. Basically, I would like it to be obvious that in the event the state is reset with a kernel/binder restart that they would likely need to start from the beginning again. This of course only matters for pages where the code cells do depend on each other.

**Comment by akhmerov** [comment]
Hm, come to think of it, just showing an error message "kernel died, please restart" might be better and easier—then it's clear that all state is lost. That would still require restart to catch the kernel dying (right now it doesn't neither).

**Comment by choldgraf** [comment]
I think this also relates to #271 which could be the place where this information exists if the kernel dies

**Comment by psychemedia** [comment]
I've wondered before about whether there is a need for a (polite) keep-alive routine, eg in the context of [RISE presentations running on Binder](https://github.com/damianavila/RISE/issues/524).

*I had one embarrassing presentation where I had a 10 minute speaking slot and a slow to launch Binder demo (which was in part demoing the availability and use of Binder). I started the Binder just before the presentation started, checked it worked, then there was a hold up in the actual start of my session. When I came to code execution RISE slides, the Binder had died, and still hadn't restarted by the end on the speaking session... For students working through a Jupyter Book in eg a 40 minute lesson, if they have to repeatedly restart (and some how, re-run all above cells from their current cell), it could ruin a class session presented using Jupyter Book.
